drivers/tee/optee/ocall.c

```diff
@@ -0,0 +1,505 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2020, Microsoft Corporation
+ */
+
+#include <linux/err.h>
+#include <linux/errno.h>
+#include <linux/slab.h>
+#include <linux/tee_drv.h>
+#include <linux/types.h>
+#include "optee_private.h"
+
+static void optee_ocall_inc_memrefs_refcount(struct tee_param *params,
+					     u32 num_params,
+					     struct optee_call_ctx *call_ctx)
+{
+	size_t n;
+
+	mutex_lock(&call_ctx->mutex);
+	for (n = 0; n < num_params; n++) {
+		struct tee_shm *shm;
+		struct tee_param *p = params + n;
+
+		if (!tee_param_is_memref(p))
+			continue;
+
+		shm = p->u.memref.shm;
+		if (!shm->ocall_link.next) {
+			list_add(&p->u.memref.shm->ocall_link,
+				 &call_ctx->list_shm);
+			tee_shm_get(p->u.memref.shm);
+		}
+	}
+	mutex_unlock(&call_ctx->mutex);
+}
+
+static void optee_ocall_dec_memrefs_refcount(struct tee_param *params,
+					     u32 num_params,
+					     struct optee_call_ctx *call_ctx)
+{
+	size_t n;
+
+	mutex_lock(&call_ctx->mutex);
+	for (n = 0; n < num_params; n++) {
+		struct tee_shm *shm;
+		struct tee_param *p = params + n;
+
+		if (!tee_param_is_memref(p))
+			continue;
+
+		shm = p->u.memref.shm;
+		if (shm->ocall_link.next) {
+			tee_shm_put(p->u.memref.shm);
+			list_del(&p->u.memref.shm->ocall_link);
+
+			/*
+			 * Set the list entry to NULL to avoid depending on
+			 * internal poison values being kept as-is in the
+			 * future. Note that this prevents the list debug
+			 * mechanism (CONFIG_DEBUG_LIST=y) from doing its job.
+			 */
+			memset(&p->u.memref.shm->ocall_link, 0,
+			       sizeof(p->u.memref.shm->ocall_link));
+		}
+	}
+	mutex_unlock(&call_ctx->mutex);
+}
+
+static void optee_ocall_cancel_worker(struct optee_call_ctx *call_ctx)
+{
+	int rc;
+	struct tee_shm *shm;
+
+	if (call_ctx->cancel_code != U32_MAX) {
+		call_ctx->rpc_arg->ret = call_ctx->cancel_code;
+		call_ctx->rpc_arg->ret_origin = TEEC_ORIGIN_COMMS;
+	}
+
+	rc = optee_do_call_with_ctx(call_ctx);
+	if (rc == -EINTR)
+		pr_warn("cancellation of OCALL was marked as cancelled");
+	else if (rc == -EAGAIN)
+		pr_warn("received an OCALL while cancelling an OCALL");
+
+	call_ctx->cancelled = true;
+
+	optee_ocall_epilogue(call_ctx);
+
+	list_for_each_entry(shm, &call_ctx->list_shm, ocall_link)
+		tee_shm_put(shm);
+
+	if (call_ctx->cancel_cb)
+		call_ctx->cancel_cb(call_ctx);
+}
+
+static void optee_ocall_finalize(struct optee_context_data *ctxdata,
+				 struct optee_call_ctx *call_ctx)
+{
+	atomic_set(&call_ctx->attached, false);
```

Comment:
I have my doubts about using `atomic_set()` and friends. I'd prefer if we could use a mutex instead.

Reply:
I'll see what I can do.

Given that this is a flag that tells the function whether to proceed, the effects of using `atomic_*` on it and protecting it with a mutex are the same as far as I can tell.